app/models/components/course/experience_points_records_ability_component.rb

```diff
@@ -0,0 +1,49 @@
+module Course::ExperiencePointsRecordsAbilityComponent
+  include AbilityHost::Component
+
+  def define_permissions
+    # cancancan is unable to merge abilities if administrators are not excluded
+    if user && !user.administrator?
+      allow_staff_manage_experience_points_records
+      allow_staff_and_points_owner_read_experience_points_records
+    end
+
+    super
+  end
+
+  private
+
+  def allow_staff_and_points_owner_read_experience_points_records
+    staff_roles = CourseUser::STAFF_ROLES.to_a.map { |role| CourseUser.roles[role] }
+    accessible_records  =
+      Course::ExperiencePointsRecord.
+      joins { course_user.course.course_users }.
+      where do
+        ((course_user.course.course_users.user_id == my { user.id }) &
+         (course_user.course.course_users.workflow_state == 'approved') &
+         (course_user.course.course_users.role >> my { staff_roles })) |
+        (course_user.user_id == my { user.id })
+      end.distinct.order(updated_at: :desc)
+
+    can :read, Course::ExperiencePointsRecord, accessible_records do |record|
+      course_user = get_course_user(record)
+      course_user && ((course_user.id == record.course_user_id) || approved_staff(course_user))
+    end
+  end
+
+  def allow_staff_manage_experience_points_records
+    can [:create, :update, :destroy], Course::ExperiencePointsRecord, [] do |record|
```

Comment:
I remember Initially we are trying to define a SQL for all the manageable records as well, that does not work ?